lib/web_ui/dev/test_runner.dart

```diff
@@ -106,6 +107,9 @@ class TestCommand extends Command<bool> {
     SupportedBrowsers.instance
       ..argParsers.forEach((t) => t.parseOptions(argResults));
 
+    // Make sure the browser requested is supported.
+    assert(SupportedBrowsers.instance.supportedBrowserNames.contains(browser));
```

Comment:
We AOT-compile the `felt` tool for speed, so asserts won't trigger. I'd recommend using a plain `if` block.

Reply:
> I was imagining that we'd have `class MobileSafari extends Browser`, just like we do for desktop Chrome, Firefox, Safari, and Edge. Is there something about mobile Safari that prevents us from using the established framework for unit-tests? In particular, I'm not sure that having to change the test code in order to be able to run on mobile is good for long-term code health.

This change is only for mobile unit tests. Other tests does not enter this code branch.

Yes, there is a reason we are totally changing our strategy here. We are not able to run the browser otherwise. One good part is though, if we ever have enough device support in the device lab to run presubmit tests on devices. Unit tests run on the real devices' browsers flawlessly using these strategy. 

I think that part we already discussed in the design doc (which I just checked you also LGTM'ed :) )

Reply:
> We AOT-compile the `felt` tool for speed, so asserts won't trigger. I'd recommend using a plain `if` block.

Thanks! I'll change the asserts.

Reply:
I can see why we want to use the driver to start/stop/manipulate the browser. That part LGTM. What I'm confused about is why we cannot implement the `Browser` interface on top of the driver and use `test_platform.dart` like we do for other browsers (i.e. hide the driver as an implementation detail). It may be the right thing to do. I'm just trying to understand the rationale. I'd like to make sure this code can scale as we fill our test matrix with more browser/OS combinations.

Reply:
(1) Can you please elaborate on the concerns? Best by providing concrete examples. For example is there a number of tests webdev will have an issue on serving? (I'm trying to understand the numbers/facts behind the scalability concern).

Reply:
(2) The way the code organized right now is:
(a) the test runner build the tests.
(b) then it registers browser as a platform using this a hack (https://github.com/dart-lang/test/blob/master/pkgs/test_core/lib/src/runner/hack_register_platform.dart)
(c) opens the browser using the code we wrote in browser files.
(d) uses the test runner to run the tests. 

On the other hand the suggested method uses:
(1) webdev serve to serve the unit tests and 
(2) uses web driver to access the browser.
(3) collect the results using the driver commands registered to each test.
I think, the overall strategy is the canonical way of using the driver (having a server, serving a file, using driver to access it, creating a communication bridge between the driver and the file under examination). I don't think we will be able to get away with using test runner in case of mobile devices. (Please let me know if you have any concrete examples running this way.)

I am also sure that this statement is not possible: _"hide the driver as an implementation detail"_, if we want to run a test using driver, we have to make sure that it responds to the questions driver ask to it (aka driver commands).

Reply:
> (1) Can you please elaborate on the concerns?

Sure:

1. We'll have a branch in `felt` that runs unit-tests using one system (`test_platform.dart`) and a second branch using entirely separate system (`mobile_unit_test_manager.dart`). This adds complexity.
1. It requires a second binding and separate code path in the test itself, the `MobileUnitTestBindings mobileUnitTestBindings = isIosSafari ? MobileUnitTestBindings() : null;` and the `setUpAll`/`tearDownAll`. This adds complexity to tests.

> I don't think we will be able to get away with using test runner in case of mobile devices.

If that's the case then we'll have to pay the cost of this complexity. I just hoped we could avoid it.

> if we want to run a test using driver, we have to make sure that it responds to the questions driver ask to it (aka driver commands).

My suggestion was not to run the test using the driver, but to use the driver to launch the browser and point it to the relevant URL. The `BrowserPlatform` class would then establish the connection with the test using the usual web unit-test protocol and run the test, like it already does for all other browsers. This way the test does not need to respond to driver commands at all.

If this is not possible, I'd be interested in knowing why. Let's sync offline. It might be easier to explain it to me in person.

Reply:
closing this comment since we made an offline meeting on this. We will keep the both branches this quarter and discuss in a few months.